Fix #21722: Add support for fetching questions in Android data handler

BenHenning opened the pull request:
## Overview

Fixes #21722

<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #[fill_in_number_here].
2. This PR does the following: [Explain here what your PR does and why]
3. (For bug-fixing PRs only) The original bug occurred because: [Explain what
   the cause of the bug was, and which PR introduced it]

## Essential Checklist

Please follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).

- [ ] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [ ] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [ ] I have written tests for my code.
- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [ ] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Testing doc (for PRs with Beam jobs that modify production server data)

<!--
If this PR affects production server data, please follow
[these instructions](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production#submitting-a-pr-with-a-new-job-or-feature-that-requires-third-party-api)
and link to the job request doc here.

Otherwise, please delete this section.
-->

- [ ] A testing doc has been written: ... (ADD LINK) ...
- [ ] _(To be confirmed by the server admin)_ All jobs in this PR have been verified on a live server, and the PR is safe to merge.

## Proof that changes are correct

<!--
Add before-and-after videos/screenshots of the user-facing interface (including
the browser devtools console) to demonstrate that the changes made in this PR
work correctly. Make sure the actions taken in the before and after videos are
the same. (For changes involving responsiveness or adjustment of UI elements,
this should be a video that gradually resizes the viewport from wide to narrow
and back again.) If this PR is for a developer-facing feature, provide
videos/screenshots of the developer-facing interface instead.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

BenHenning committed:
Add support for fetching questions.

This introduces basic support for introducing questions from the Android
data controller. Manual and automated testing is still needed.

BenHenning commented:
I'll fill in the description later once I actually finish the PR. In the meantime, @seanlip it'd be nice to get your or someone else's quick thoughts on the approach here. I realized that the existing ``question_fetchers`` functionality won't quite work here, so I split up the fetch into two steps:
1. Fetch all of the question IDs associated with a given skill ID.
2. Fetch the question corresponding to a specific skill ID.

This should actually play more nicely with the parallelization of the script, too, though it does put a lot of load on the server. I'm wondering slightly if we should try to introduce get_multi with version support here (do the versioned models even support that?).

Separately, I'm sure you're not keen on the direct model access in the data controller. I suspect that I should introduce another method in ``question_fetchers`` to fetch the link information?

seanlip commented on core/controllers/android.py:
```diff
@@ -231,6 +240,33 @@ def get(self) -> None:
+        elif activity_type == 'question_skill_link':
+            for activity_data in activities_data:
+                question_skill_link_models = (
+                    question_models.QuestionSkillLinkModel
```
This shouldn't directly reference the storage layer, yeah. 

Re load, thoughts on using question_fetchers.get_questions_and_skill_descriptions_by_skill_ids() and passing in multiple skill IDs at a time? Otherwise maybe you'll need to make a new function in question_fetchers for this use case.

BenHenning replied:
Yeah, the latter is a good idea and what I've implemented in the latest version. I also introduced new ``question_fetchers`` functionality to avoid the direct storage model reference here.

seanlip commented on core/domain/question_fetchers.py:
```diff
@@ -86,6 +86,32 @@ def get_questions_and_skill_descriptions_by_skill_ids(
     return questions, grouped_skill_descriptions
 
 
+def get_question_by_id(
```
I think you can just call get_questions_by_id() with a single-element list (or use that existing function directly where you need it).

BenHenning replied:
I don't think that's correct since we need to be able to fetch specific versions. That being said, I think there's some additional cleanup possible here after the latest changes.

seanlip replied:
Ah yes, sorry -- I missed that you needed versions. I think with the new function you have below, you can probably drop this one, yeah.

seanlip commented:
Thanks @BenHenning -- left some comments above.

What's the API you're thinking of for "introduce get_multi with version support here (do the versioned models even support that?)"? Are you passing in a list of (model ID, version) tuples? If so then it might be possible to handle that but it might require updates to base_models.VersionedModel class to support a get_multi() at the storage layer.

BenHenning committed:
Make an attempt at improving fetching.

This commit is actually doing a bunch of stuff, including:
- Introducing first-class functionality in question_fetchers for
  fetching the skill<->question ID links.
- Introducing base VersionedModel support for fetching multiple entities
  by IDs and versions (rather than just multiple versions with one ID).
- Updating all activity fetchers used by the Android data controller to
  utilize this new VersionedModel functionality to support fetching
  multiple activities by IDs and versions.
- Updating the Android data controller to use these new fetcher methods
  over fetching structures one-by-one.
- Incidentally stop using Redis for Android data controller caching due
  to the new methods not checking for cached structures.

This still requires substantial reformatting and documentation work, in
addition to actual manual and automated testing. However, it's expected
that this will provide substantial performance and stability
improvements to the data controller since the frontend in this case (the
Android lesson script) was updated several months back to now leverage
multi-version fetching. There's also substantially less risk to
introducing an outage due to Redis filling up from the large number of
requests that the script can sometimes send during a full download.

BenHenning commented:
> Thanks @BenHenning -- left some comments above.
> 
> What's the API you're thinking of for "introduce get_multi with version support here (do the versioned models even support that?)"? Are you passing in a list of (model ID, version) tuples? If so then it might be possible to handle that but it might require updates to base_models.VersionedModel class to support a get_multi() at the storage layer -- perhaps worth filing an issue and saving that for a separate PR if it's needed.

As discussed, it's a bit of work but probably worth introducing since I think we stand to gain fairly substantial improvements. PTAL at the latest changes @seanlip and let me know your thoughts (please ignore the fact that things are probably super broken; still haven't actually tested it yet, just looking for design feedback and suggestions).

Incidentally, the latest changes actually fix the 'Redis' part of #21253 since I introduced entirely new fetcher methods which don't bother interacting with the Redis cache. This means much less risk to bring down the production environment when running the script. I'm also hoping that all of the placed ``get_multi``s will improve performance (or at least reduce datastore costs).

From a cherry-pick perspective, this change is probably going to be 1k-1.5k LOC by the time it's done (maybe even close to 2k depending on how much setup work all the potential tests will need). However, the only surface being changed that could affect existing behavior is ``VersionedModel``'s reconstitution logic, and that seems extremely safe to me (and massively tested by all of the backend tests that interact with versioned models).

BenHenning commented:
Also separately @seanlip Oppiabot seems a bit spammy with failures--is that WAI?

seanlip commented on core/controllers/android.py:
```diff
@@ -255,32 +238,63 @@ def get(self) -> None:
-                    'payload': (
-                        topic.to_dict() if topic is not None else None)
-                })
+            # Otherwise it's a standard versioned activity that requires a
```
Maybe worth leaving a similar comment in each branch of this if/else explaining why we choose the approach given for that entity (i.e. why doesn't it count as a "standard versioned activity").

seanlip commented on core/domain/question_fetchers.py:
```diff
@@ -108,6 +134,31 @@ def get_questions_by_ids(
+        if question_model is not None else None]
+
+
+def get_question_ids_by_skill_ids(skill_ids: List[str]) -> Dict[str, List[str]]:
```
Not sure if this will give you a balanced list of questions. AFAICT the get_question_skill_links_by_skill_ids() method sorts everything by last_updated and returns the most recent data. So if we create all the questions for one skill first and then create all the questions for another, you might only get the second one.

You might need another method in the storage layer to handle the query you want in a more balanced way (fetching at most X questions for each skill), if it doesn't already exist. Not really sure how to do that without making one query for each skill ID, though.

BenHenning replied:
The goal is not to fetch a balanced list of questions though, the goal is to fetch *all* of them. The action of compiling questions into a practice session is the responsibility of the app, so we need the entire set of questions downloaded in order to do that effectively.

seanlip replied:
Ah ok, if you actually want all, then you might probably need a paged approach since I believe we already have > 1000 questions. (And in any case, you might need to shard into bundles of e.g. 20 skills, anyway -- if you look at the impl of get_question_skill_links_by_skill_ids() it uses an ndb "IN" filter and I've never tested it with 100+ items at a time, so I'm not sure what's the max number of skill IDs it will support.)

OK with testing it out as-is if you like but I suspect there is some hidden limit.

seanlip replied:
Per conversation with Ben -- paged approach is OK.

seanlip commented on core/domain/subtopic_page_services.py:
```diff
@@ -207,6 +207,23 @@ def get_subtopic_pages_with_ids(
+) -> List[Optional[subtopic_page_domain.SubtopicPage]]:
+    subtopic_page_ids_and_versions = [
+        (subtopic_page_domain.SubtopicPage.get_subtopic_page_id(
+            topic_id, subtopic_id), version)
```
subtopic_index?

BenHenning replied:
The backend is a bit inconsistent with this, so I wasn't sure if I should continue that term outside of android.py (or if I should even use it at all). Sometimes it's subtopic_id, subtopic_index, or subtopic_page_id.

seanlip replied:
Ah -- there are two different things here, not sure which you want.

subtopic_id is a unique ID for the subtopic. (Also called subtopic_page_id, but the domain object in topic_domain.py is Subtopic so subtopic_id is the more correct term.)
subtopic_index is the index of that subtopic in the list of subtopics. See get_subtopic_index() in topic_domain.py.

Not sure which of those you actually want. If subtopic_id then this probably makes sense to keep as subtopic_id but then I'm not sure if the arg to this function is correct.

seanlip commented on core/controllers/android.py:
```diff
@@ -178,59 +181,41 @@ def get(self) -> None:
+            split_ids_versions = [
+                (activity_data['id'].split('-'), activity_data.get('version'))
+                for activity_data in activities_data]
+            reference_ids_versions = [
```
This sounds like 2-tuples, probably want a name that conveys 3-tuples

seanlip commented on core/domain/translation_fetchers.py:
```diff
@@ -33,6 +33,15 @@
     models.Names.TRANSLATION])
 
 
+class EntityTranslationReferenceDict(TypedDict):
```
Probably better to drop this and reuse the one from core/storage

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -1496,6 +1501,46 @@ def get_version(
+    @classmethod
+    def get_version_multi(
+        cls: Type[SELF_VERSIONED_MODEL],
+        entity_ids_versions: List[Tuple[str, Optional[int]]]
```
entity_ids_and_versions

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -1496,6 +1501,46 @@ def get_version(
+        # The assumption is that every instance of cls uses the same
+        # SNAPSHOT_CONTENT_CLASS, so only one fake instance is needed to fetch
+        # that class for a get_multi() operation.
+        temp_model = cls(id='', version_number=0)
```
How come we can't just do cls.SNAPSHOT_CONTENT_CLASS? Not sure why we need a temp model here.

BenHenning replied:
Uh, my Python knowledge has greatly degraded over the years. :) I assumed that this was an instance property, not a class property, as other uses are also referencing it from a member rather than the class. Is that not correct?

seanlip replied:
It's defined as a class property in VersionedModel so I believe you can access it from the class as well, similar to COMMIT_TYPE_CHOICES.

seanlip commented on core/storage/base_model/gae_models.py:
```diff
@@ -1496,6 +1501,46 @@ def get_version(
+            cls.get_snapshot_id(
+                entity_id,
+                version_number
+                if version_number else latest_versions_by_id[entity_id])
```
The optionality here feels odd. We can't make the version number a required input?

(I would generally prefer to keep the core data-fetching functionality simple and easy to audit.)

BenHenning replied:
This slightly matches other functionality already in VersionedModel (in that get can take a ``None`` version if I recall). Usually 'None' refers to 'latest' as we don't have a better sentinel value to use for that case. I could introduce a separate API for latest, but that gets a bit messy as it would require two calls (and thus two ``get_multi``s) unless we put the onus on the caller to compile the snapshot IDs.

Do you have a better suggestion?

seanlip replied:
Sorry, my question was unclear. I don't really understand how the controller API is being used, so I wanted to check why the app needs both 'latest' and 'per-version'. Is the latter alone not sufficient?

If not then what you have here is probably the best approach but I wanted to double-check what is actually needed (and why).

seanlip commented on core/storage/translation/gae_models.py:
```diff
@@ -131,6 +140,21 @@ def get_model(
+    @classmethod
+    def get_model_multi(
+        cls,
+        entity_configs: List[EntityTranslationReferenceDict]
```
entity_translation_references (let's standardize terminology)

seanlip reviewed (changes requested):
Thanks @BenHenning! Took a look. Overall I think the storage-layer changes seem OK, though I left some comments, see what you think. 

Re CP-ing, yeah it's hard to say. Let's cross that bridge when we come to it. If you want to test then here's the head branch of the upcoming release that you can check against (since it's where a hotfix would likely land): https://github.com/oppia/oppia/tree/release-3.4.4
